### DISCUSSION TITLE
What is the best way to support multiple (in hundreds or even thousands) request to trafilatura?

### DISCUSSION THREAD

**Discussion by krstp** [adbar/trafilatura#515] [General]
Currently Trafilatura works in much synchronous way due to underlying `urllib` dependency. What I am finding when reaching approx. 60-80 req/sec the engine somewhat locks out. I wonder how people manage multiple extraction requests that work in expected concurrent and efficient way?

PS. I am not finding any existing conversations on the subject.
PS2. I see there is Trafilatura API that supports large volume requests... I wonder how the backend solution got handled here in terms of underlying `urllib` 🤔  Ref: https://rapidapi.com/trafapi/api/trafilatura

What is interesting, based on the "large volume"[ rapidAPI endpoint Trafilatura hosts](https://rapidapi.com/trafapi/api/trafilatura), the latency on single request is approx `6000ms`... which is `6sec`, this is not blazing fast turnout in terms of high volume of data extraction (see screenshot). Note, this is not a complain on the Trafilatura itself, but rather backend solution for multiple and concurrent (perfectly in fully `async` way, i.e. `aiohttp/asyncio`) requests handling. I less care how much time the extraction takes or full extraction and return loop, could be `1sec`, could be `30sec`, I would rather prefer to have good support for concurrent requests, however, from what I see atm Trafilatura works in much synchronous way.

![Screenshot 2024-02-26 at 9 15 39 AM](https://github.com/adbar/trafilatura/assets/4711498/f5ad3494-38b1-4ef9-8fae-6f398375ccce)

Ref:
- related issue: https://github.com/adbar/trafilatura/issues/514

**Comment by adbar** [comment]
Hi @krstp, there has been a problem with the latency, I'm investigating, it was between 0,5 and 1 sec before that.

The server is lean and Trafilatura distributes requests among hosts to allow for fluid downloads and also to avoid being blocked.

**Reply by adbar** [reply]
I checked, for the record: if a web page is slow to respond, then the API response will be slow to come but it has nothing to do with the API itself. So the statistics are not quite relevant to assess the performance of the package in itself here.
